### PR TITLE
Install files using CMake, use CMake for Nix

### LINKED ISSUE #4956: Add cmake install target
### Description

Currently it seems like the makefile is used for the install target even though ninja is used via cmake in the makefile.
I think it would make more sense to add a cmake install target so that you can do this:
```
cmake -B build -G Ninja
cmake --build build
sudo cmake --install build
```

This would also benefit the use case of customizing cmake features such as `-DNO_XWAYLAND=ON` or `-DCMAKE_BUILD_TYPE=MinSizeRel` without having to modify the makefile. This also impacts setting custom compilers and/or compiler flags.

### PR DESCRIPTION
**Pull request by fufexan**
#### Describe your PR, what does it fix/add?

CMake now installs files directly (through `cmake --install ./build`). The Makefile commands have a warning that directs users to the appropriate commands.

Nix now uses CMake for building. We also no longer patch the build process to avoid building subprojects.
This simplifies things on the building part, but complicates them a bit on the UX part.

In order to test these changes, you will have to do so with the following command:
```console
$ nix build 'git+https://github.com/hyprwm/Hyprland?ref=cmake&submodules=1'
```

Closes #4956.
Closes #2313.

> [!IMPORTANT]
> I may revert back to patching the build process and using derivations for subprojects instead, if this method proves to be too unergonomic.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

CMake run from the command line seems to build fine, but when built with Nix it seems to build twice for some reason. Will have to look into the CMake and Ninja setup hooks.

Another thing to note is, in Nix at least, the XWayland package contains a pkg-config file that exposes `includedir` and `Cflags`, which are useless due to XWayland not including headers in the package. They have to be patched out.

#### Is it ready for merging, or does it need work?

No. Please test it and report back.

TODO:
- [x] Fix Nix build compiling twice (was due to `CONFIGURE_DEPENDS`)
- [x] Check feature parity with the Meson build
- [x] Nix devshell does not inherit overridden XWayland (you have to build using `cmakeConfigurePhase`, `ninjaBuildPhase` and `ninjaInstallPhase`)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -26,81 +26,24 @@ clear:
...
-	find src -name '*.h*' -print0 | cpio --quiet -0dump ${PREFIX}/include/hyprland
-	cd subprojects/wlroots-hyprland/include && find . -name '*.h*' -print0 | cpio --quiet -0dump ${PREFIX}/include/hyprland/wlroots-hyprland && cd ../../..
-	cd subprojects/wlroots-hyprland/build/include && find . -name '*.h*' -print0 | cpio --quiet -0dump ${PREFIX}/include/hyprland/wlroots-hyprland && cd ../../../..
-	cp ./protocols/*-protocol.h ${PREFIX}/include/hyprland/protocols
-	cp ./build/hyprland.pc ${PREFIX}/share/pkgconfig
-	if [ -d /usr/share/pkgconfig ]; then cp ./build/hyprland.pc /usr/share/pkgconfig 2>/dev/null || true; fi
-
-	chmod -R 755 ${PREFIX}/include/hyprland
-	chmod 755 ${PREFIX}/share/pkgconfig
+	@echo -en "$(MAKE) installheaders has been deprecated.\nHeaders are now installed by default using CMake\n"
```
Comment (vaxerski) [inline Makefile:45]: make installheaders is used by hyprpm

File: Makefile [hunk 2]
```diff
@@ -26,81 +26,24 @@ clear:
...
-	if [ ! -f ${PREFIX}/share/wayland-sessions/hyprland.desktop ]; then cp ./example/hyprland.desktop ${PREFIX}/share/wayland-sessions; fi
-	cp ./assets/wall* ${PREFIX}/share/hyprland
-	mkdir -p ${PREFIX}/share/xdg-desktop-portal
-	cp ./assets/hyprland-portals.conf ${PREFIX}/share/xdg-desktop-portal
-
-	mkdir -p ${PREFIX}/share/man/man1
-	install -m644 ./docs/*.1 ${PREFIX}/share/man/man1
-
-	$(MAKE) installheaders
+	@echo -en "$(MAKE) release has been deprecated.\nYou can install Hyprland using $(MAKE) all\n"
```
Comment (vaxerski) [inline Makefile:33]: typo release -> install

File: Makefile (hunk 1 above)
Comment (fufexan) [reply Makefile:45]: Alright I'll leave that there then.

File: CMakeLists.txt [hunk 3]
```diff
@@ -269,3 +270,50 @@ protocolNew("unstable/xdg-output/xdg-output-unstable-v1.xml" "xdg-output-unstabl
...
+        DESTINATION ${CMAKE_INSTALL_DATADIR}/xdg-desktop-portal)
+
+# man pages
+file(GLOB_RECURSE MANPAGES CONFIGURE_DEPENDS "docs/*.1")
+install(FILES ${MANPAGES}
+        DESTINATION ${CMAKE_INSTALL_MANDIR}/man1)
+
+
+# headers
+file(GLOB_RECURSE HEADERS_HL CONFIGURE_DEPENDS "*.h*")
```
Comment (vaxerski) [inline CMakeLists.txt:308]: shouldn't this be `src/*.h*`? Otherwise it will consume all the subprojects and shit

File: Makefile [hunk 4]
```diff
@@ -26,72 +26,30 @@ clear:
...
-	rm -f ${PREFIX}/share/zsh/site-functions/_hyprpm
+	@echo -en "$(MAKE) uninstall has been deprecated.\nYou can uninstall Hyprland using CMake\n"
+	@exit 1
 
 pluginenv:
 	@echo -en "$(MAKE) pluginenv has been deprecated.\nPlease run $(MAKE) all && sudo $(MAKE) installheaders\n"
 	@exit 1
 	
-installheaders:
+installheader:
```
Comment (vaxerski) [inline Makefile:44]: unintentional change?

File: CMakeLists.txt (hunk 3 above)
Comment (fufexan) [reply CMakeLists.txt:308]: Yup, missed that.

File: hyprctl/CMakeLists.txt [hunk 5]
```diff
@@ -5,4 +5,17 @@ project(
...
+install(TARGETS hyprctl)
+
+# shell completions
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/hyprctl.bash
+        DESTINATION ${CMAKE_INSTALL_DATADIR}/bash-completion/completions
+        RENAME hyprctl)
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/hyprctl.fish
+        DESTINATION ${CMAKE_INSTALL_DATADIR}/fish/vendor_completions.d)
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/hyprctl.zsh
+        DESTINATION ${CMAKE_INSTALL_DATADIR}/zsh/site_functions
```
Comment (jbeich) [inline hyprctl/CMakeLists.txt:20]: Typo? s/_/-/ to match Makefile (and Meson):
```suggestion
        DESTINATION ${CMAKE_INSTALL_DATADIR}/zsh/site-functions
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -282,3 +283,58 @@ protocolNew("staging/ext-session-lock/ext-session-lock-v1.xml" "ext-session-lock
...
+
+# man pages
+file(GLOB_RECURSE MANPAGES "docs/*.1")
+install(FILES ${MANPAGES}
+        DESTINATION ${CMAKE_INSTALL_MANDIR}/man1)
+
+
+# pkgconfig entry
+install(FILES ${CMAKE_BINARY_DIR}/hyprland.pc
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
```
Comment (jbeich) [inline CMakeLists.txt:322]: Typo? s/lib/share/ to match Makefile (and Meson):
```suggestion
        DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig)
```

File: Makefile [hunk 7]
```diff
@@ -30,58 +30,16 @@ clear:
...
-	if [ ! -f ${PREFIX}/share/wayland-sessions/hyprland.desktop ]; then cp ./example/hyprland.desktop ${PREFIX}/share/wayland-sessions; fi
-	cp ./assets/wall* ${PREFIX}/share/hyprland
-	mkdir -p ${PREFIX}/share/xdg-desktop-portal
-	cp ./assets/hyprland-portals.conf ${PREFIX}/share/xdg-desktop-portal
-
-	mkdir -p ${PREFIX}/share/man/man1
-	install -m644 ./docs/*.1 ${PREFIX}/share/man/man1
-
-	$(MAKE) installheaders
+	@echo -en "$(MAKE) install has been deprecated.\nYou can install Hyprland using $(MAKE) all\n"
```
Comment (vaxerski) [inline Makefile:37]: make all doesn't install. make install should just run `cmake --install build`.

File: Makefile [hunk 8]
```diff
@@ -30,58 +30,16 @@ clear:
...
-	rm -rf ${PREFIX}/share/hyprland
-	rm -f ${PREFIX}/share/man/man1/Hyprland.1
-	rm -f ${PREFIX}/share/man/man1/hyprctl.1
-	rm -f ${PREFIX}/share/bash-completion/completions/hyprctl
-	rm -f ${PREFIX}/share/fish/vendor_completions.d/hyprctl.fish
-	rm -f ${PREFIX}/share/zsh/site-functions/_hyprctl
-	rm -f ${PREFIX}/share/bash-completion/completions/hyprpm
-	rm -f ${PREFIX}/share/fish/vendor_completions.d/hyprpm.fish
-	rm -f ${PREFIX}/share/zsh/site-functions/_hyprpm
+	@echo -en "$(MAKE) uninstall has been deprecated.\nYou can uninstall Hyprland using CMake\n"
```
Comment (vaxerski) [inline Makefile:41]: same shit

File: hyprland.pc.in [hunk 9]
```diff
@@ -1,8 +1,8 @@
...
-includedir="${prefix}/include"
+prefix=@PREFIX@
+includedir=@INCLUDEDIR@
 
 Name: Hyprland
 URL: https://github.com/hyprwm/Hyprland
 Description: Hyprland header files
 Version: @HYPRLAND_VERSION@
-Cflags: -I"${includedir}/hyprland/protocols" -I"${includedir}/hyprland/wlroots-hyprland" -I"${includedir}"
+Cflags: -I${includedir} -I${includedir}/hyprland/protocols -I${includedir}/hyprland
```
Comment (vaxerski) [inline hyprland.pc.in:8]: why was the wlroots includedir nuked?

Comment (fufexan) [reply hyprland.pc.in:8]: Wasn't nuked, it still works. See the related plugins PR. It just didn't work unless I put the include dir in the parent dir.

Comment (vaxerski) [reply hyprland.pc.in:8]: uh huh. ok

**Comment by fufexan** [comment]
@vaxerski I've also removed all occurrences of `-j`, as per https://stackoverflow.com/questions/70541788/how-to-build-a-cmake-project-in-parallel-on-all-available-cores.

**Comment by fufexan** [comment]
@outfoxxed can you test when you have the time?

**Comment by outfoxxed** [comment]
Headers seem to work now, hy3 built successfully.

**Review by vaxerski — APPROVED** [review]
vă mulțumesc pentru munca frate

**Pull request merged**